visualizer/semanticsHelper.js

```diff
@@ -35,7 +182,99 @@
+  });
+
+  function populateSemanticsResult(traceNode, optOpName, optArguments) {
+    var operationName = optOpName || opName;
+    try {
+      var nodeWrapper = semantics._getSemantics().wrap(traceNode.bindings[0]);
+      if (operationName in semantics._getSemantics().operations) {
+        var argValues;
+        if (operationName === opName) {
+          argValues = toValueList(opArguments || Object.create(null));
+        } else {
+          argValues = toValueList(optArguments || Object.create(null));
+        }
+        nodeWrapper[operationName].apply(nodeWrapper, argValues);
+      } else {
+        nodeWrapper._forgetMemoizedResultFor(operationName);
+        nodeWrapper[operationName];   // eslint-disable-line no-unused-expressions
+      }
+    } catch (error) {
+      /* All the error will be an ErrorWrapper which already recorded in the resultMap */
+    }
+  }
+  ohmEditor.parseTree.addListener('render:parseTree', function(traceNode) {
+    if (!opName) {
+      ohmEditor.semantics.appendEditor = false;
+      return;
+    }
+    ohmEditor.semantics.appendEditor = true;
+    initializeSemanticsLog();
+    populateSemanticsResult(traceNode);
+  });
+
+  function forceResult(traceNode) {
+    forcing = true;
+    var semanticsNameList = semantics.getOperationNames().concat(semantics.getAttributeNames());
+    semanticsNameList.forEach(function(name) {
+      populateSemanticsResult(traceNode, name);
+    });
+    forcing = false;
+  }
+
+  function getDefaultArgExpression(traceNode) {
+    var pexpr = traceNode.expr;
+
+    // Get rule body of the Apply expression.
+    if (pexpr instanceof ohm.pexprs.Apply) {
+      var succeedChild = traceNode.children[traceNode.children.length - 1];
+      pexpr = succeedChild.expr;
+
+      // If the rule body is an Alt expression, then get its succeed term.
+      if (pexpr instanceof ohm.pexprs.Alt) {
+        pexpr = succeedChild.children[succeedChild.children.length - 1].expr;
+      }
+    }
+
+    return pexpr;
   }
-  ohmEditor.semantics.addListener('add:semanticOperation', addSemanticOperation);
+
+  // Exports
+  // -------
+
+  ohmEditor.semantics.getResults = function(traceNode) {
+    var key = nodeKey(traceNode.bindings[0]);
+    if (!(key in resultMap)) {
+      forceResult(traceNode);
+    }
+    return resultMap[key];
+  };
+
+  // If there is no user added action for the rule, return default argument list,
+  // Otherwise, return the argument list that user renamed before
+  ohmEditor.semantics.getActionArgPariedList = function(traceNode) {
+    var actionKey = traceNode.bindings[0].ctorName;
+    var defaultArgExpression = getDefaultArgExpression(traceNode);
+
+    var argPairList = {argExpr: defaultArgExpression};
+    var action = semantics._getActionDict(opName)[actionKey];
+    if (!action) {
+      return argPairList;
+    }
+
+    var actionStr = action.toString();
+    var isDefault = actionStr.indexOf('/* _DEFAULT_ */') + 15 === actionStr.indexOf('\n');
```

Comment:
This is very fragile -- it's not obvious from the functions above that the comment `/* _DEFAULT_ */` is there for a reason. And in general, it's not a good idea to use string comparison to determine function identity -- there's too many subtle things that can go wrong.

What about just putting a known property on those functions? E.g., in `initActionDict`, you could do something like this:

```
var defaults = {
  _terminal: function() { ... },
  _iter: function(children) { ... },
  _nonterminal: function(children) { ... },
};
defaults.forEach(function(name, fn) {
  fn._isDefault = true;
});
```

Then here you could just check `action._isDefault`.